engines/cryomni3d/dialogs_manager.cpp

```diff
@@ -0,0 +1,558 @@
+			// IF is not taken, go to next line
+			return false;
+		}
+
+		ifLine = testValue + 1;
+		for (; *ifLine == ' ' || *ifLine == '\t'; ifLine++) { }
+
+		if (!strncmp(ifLine, "AND IF ", 7)) {
+			ifLine += 7;
+		} else {
+			finishedConditions = true;
+		}
+	}
+
+	/* We are in the (implicit) THEN part of the IF
+	 * ifLine points to the instruction */
+	if (!strncmp(ifLine, "GOTO", 4)) {
+		buildGotoGoto(ifLine, gotos);
+	} else if (!strncmp(ifLine, "LET", 3)) {
+		executeLet(ifLine);
+	} else if (!strncmp(ifLine, "SHOW", 4)) {
+		executeShow(ifLine);
+	} else {
+		debug("Invalid IF line: %s", ifLine);
+		return false;
+	}
+
+	return true;
+}
+
+void DialogsManager::executeLet(const char *letLine) {
+	letLine = letLine + 4;
+
+	const char *endVar = letLine;
+	const char *equalPos;
+	// Find next '='
+	for (; *endVar != '='; endVar++) { }
+	equalPos = endVar;
+	// Strip spaces at the end
+	endVar--;
+	for (; *endVar == ' '; endVar--) { }
+	endVar++;
+	Common::String variable(letLine, endVar);
+
+	(*this)[variable] = equalPos[1];
+}
+
+void DialogsManager::executeShow(const char *showLine) {
+	showLine = showLine + 5;
+
+	const char *endShow = showLine;
+	// Find next ')' and include it
+	for (; *endShow != ')'; endShow++) { }
+	endShow++;
+
+	Common::String show(showLine, endShow);
+
+	executeShow(show);
+}
+
+const char *DialogsManager::executePlayerQuestion(const char *text, bool dryRun,
+        const char **realLabel) {
+	// Go after the text
+	const char *actions = nextLine(text);
+
+	while (actions && *actions != ':') {
+		if (!strncmp(actions, "IF ", 3)) {
+			actions = parseIf(actions);
+		} else if (!strncmp(actions, "LET ", 4)) {
+			if (!dryRun) {
+				executeLet(actions);
+			}
+			actions = nextLine(actions);
+		} else if (!strncmp(actions, "GOTO ", 5)) {
+			return findLabel(actions + 5, realLabel);
+		} else {
+			actions = nextLine(actions);
+		}
+	}
+
+	// There were no GOTO, so it's the end of the conversation
+	return nullptr;
+}
+
+const char *DialogsManager::parseIf(const char *ifLine) {
+	ifLine += 3;
+
+	bool finishedConditions = false;
+	while (!finishedConditions) {
+		const char *endVar = ifLine;
+		const char *equalPos;
+		// Find next '='
+		for (; *endVar != '='; endVar++) { }
+		equalPos = endVar;
+		// Strip spaces at the end
+		endVar--;
+		for (; *endVar == ' '; endVar--) { }
+		endVar++;
+		Common::String variable(ifLine, endVar);
+
+		const char *testValue = equalPos + 1;
+		for (; *testValue == ' ' || *testValue == '\t'; testValue++) { }
+
+		byte value = (*this)[variable];
+
+		if (value != *testValue) {
+			// IF is not taken, go to next line
+			return nextLine(ifLine);
+		}
+
+		ifLine = testValue + 1;
+		for (; *ifLine == ' ' || *ifLine == '\t'; ifLine++) { }
+
+		if (!strncmp(ifLine, "AND IF ", 7)) {
+			ifLine += 7;
+		} else {
+			finishedConditions = true;
+		}
+	}
+
+	/* We are in the (implicit) THEN part of the IF
+	 * ifLine points to the instruction */
+	return ifLine;
```

Comment:
I wanted to implement the same algorithms there was in the game to be sure it would behave the same. But I agree that could be done in lex/yacc.
I didn't know this was allowed in scummvm.

Reply:
Heh: https://github.com/scummvm/scummvm/blob/master/engines/director/lingo/lingo-gr.y

Reply:
Should I move to lex/yacc or can I let it like it is?
The main advantage of the current situation is that it works and is compliant with the original game.

Reply:
You can let it as is, especially since it has been debugged. It was just an open question from me.